File: 05-java-mqtt/src/main/java/org/protelis/demo/ProtelisConfigSpec.java
package org.protelis.demo;

import com.uchuhimo.konf.ConfigSpec;
import com.uchuhimo.konf.RequiredItem;
import org.protelis.demo.data.MqttProtelisNode;

import java.util.List;

/**
 * Konf specification for the TOML configuration file.
 */
@SuppressWarnings("PMD.FieldNamingConventions")
public final class ProtelisConfigSpec {

    /**
     * Create a new config specification named protelis.
     */
    public static final ConfigSpec SPEC = new ConfigSpec("protelis");
    /**
     * Required field iterations. It represents the number of cycles the simulation will perform.
     */
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<Integer> iterations =
            new RequiredItem<Integer>(SPEC, "iterations") { };
    /**
     * Required field protelisModuleName. It is the name of the .pt source file.
     */
    public static final RequiredItem<String> protelisModuleName =
            new RequiredItem<String>(SPEC, "protelisModuleName") { };

    /**
     * Required field brokerHost. It is the IP address of the MQTT broker.
     */
    public static final RequiredItem<String> brokerHost =
            new RequiredItem<String>(SPEC, "brokerHost") { };

    /**
     * Required field brokerPort. It is the port of the MQTT broker.
     */
    //@SuppressWarnings("PMD.FieldNamingConventions")
    // CHECKSTYLE: ConstantName OFF
    public static final RequiredItem<Integer> brokerPort =
            new RequiredItem<Integer>(SPEC, "brokerPort") { };

    /**
     * Required field nodes. It contains a list of protelis nodes.
     */
    public static final RequiredItem<List<MqttProtelisNode>> nodes =
            new RequiredItem<List<MqttProtelisNode>>(SPEC, "nodes") { };

    private ProtelisConfigSpec() { }
}

File: 06-kotlin-mqtt/src/main/kotlin/ProtelisConfigSpec.kt
package demo

import com.uchuhimo.konf.ConfigSpec

/**
 * This is used by Konf to map the TOML objects ProtelisNode to a class.
 */
data class ProtelisNode(val id: Int, val listen: String, val neighbors: Set<String>, val leader: Boolean = false)

/**
 * This is used by Konf to map the TOML main configuration to a class.
 */
object ProtelisConfigSpec : ConfigSpec("protelis") {
    val iterations by required<Int>(description = "Number of iterations the nodes have to run.")
    val protelisModuleName by required<String>()
    val brokerHost by required<String>()
    val brokerPort by required<Int>()
    val nodes by required<List<ProtelisNode>>(description = "List of nodes")
}
